### DISCUSSION TITLE
Rule: Subcribe a multisensor topic

### DISCUSSION THREAD

**Discussion by GVollaro** [arendst/Tasmota#11636] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

I have a device that publishes the following topic:
hasp/termo_display/state/sensors => {"Temp":"23.08","Hum":"30.80","Pres":"980.65","PresSL":"1025.88"}

I defined the following rule to generate MQTT events on publication:
Rule1 on mqtt#connected do backlog 
Subscribe DisplayTemp, hasp/termo_display/state/sensors, Temp;
Subscribe DisplayHum, hasp/termo_display/state/sensors, Hum; 
Subscribe DisplayPres, hasp/termo_display/state/sensors, PresSL endon .....

Another rule allows me to publish the intercepted values ​​on an OpenHASP display:
Rule2 on Event#DisplayTemp do publish hasp/termo_display/command/p[1].b[20].txt %value% endon
on Event#DisplayHum do publish hasp/termo_display/command/p[1].b[21].txt %value% endon
on Event#DisplayPres do publish hasp/termo_display/command/p[1].b[23].txt %value% endon ......

The problem encountered is that only the first subscribed parameter generates the relative publish event, while the next two are ignored.
Swapping the order of definition in the first rule also changes the published parameter which is always the first in the order of definition

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): NodeMCU8266
- [x] Tasmota binary firmware version number used: Tasmota 9.2.0.4
  - [ ] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: Tasmotizer
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
 
```lua
  Configuration output here:
CMD: Backlog Template; Module; GPIO 255
15:10:58.847 SRC: WebConsole from 192.168.10.11
15:10:58.850 CMD: Gruppo 0, Indice 1, Comando "BACKLOG", Dati "Template; Module; GPIO 255"
15:10:58.981 SRC: Backlog
15:10:58.984 CMD: Gruppo 0, Indice 1, Comando "TEMPLATE", Dati ""
15:10:58.991 MQT: stat/OpenTherm1/RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,4960,4928,1,1,224,1,1,1,1,1],"FLAG":0,"BASE":18}
15:10:59.203 SRC: Backlog
15:10:59.206 CMD: Gruppo 0, Indice 1, Comando "MODULE", Dati ""
15:10:59.213 MQT: stat/OpenTherm1/RESULT = {"Module":{"0":"Generic"}}
15:10:59.454 SRC: Backlog
15:10:59.457 CMD: Gruppo 0, Indice 1, Comando "GPIO", Dati "255"
15:10:59.468 MQT: stat/OpenTherm1/RESULT = {"GPIO0":{"0":"Nessuno"},"GPIO1":{"0":"Nessuno"},"GPIO2":{"0":"Nessuno"},"GPIO3":{"0":"Nessuno"},"GPIO4":{"4960":"OpenTherm - TX"},"GPIO5":{"4928":"OpenTherm - RX"},"GPIO9":{"0":"Nessuno"},"GPIO10":{"0":"Nessuno"},"GPIO12":{"224":"Relè1"},"GPIO13":{"0":"Nessuno"},"GPIO14":{"0":"Nessuno"},"GPIO15":{"0":"Nessuno"},"GPIO16":{"0":"Nessuno"},"GPIO17":{"0":"Nessuno"}}

```
- [x] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
CMD: Backlog Rule1; Rule2; Rule3
14:50:32.943 RUL: Rule1 on mqtt#connected do backlog  Subscribe DisplayTemp, hasp/termo_display/state/sensors, Temp; Subscribe DisplayHum, ha...
14:50:32.948 RUL: Rule1 +Subscribe AcquaRiscSet, hasp/termo_display/state/p2b5, text; Subscribe AcquaSanSet, hasp/termo_display/state/p2b3, t...
14:50:32.965 MQT: stat/OpenTherm1/RESULT = {"Rule1":"ON","Once":"OFF","StopOnError":"OFF","Length":751,"Free":269,"Rules":"on mqtt#connected do backlog  Subscribe DisplayTemp, hasp/termo_display/state/sensors, Temp; Subscribe DisplayHum, hasp/termo_display/state/sensors, Hum;  Subscribe DisplayPres, hasp/termo_display/state/sensors, PresSL; Subscribe DisplayPag, hasp/termo_display/state/page; Subscribe DisplayDim, hasp/termo_display/state/dim; Subscribe PulsanteAcc, hasp/termo_display/state/p1b14, val; Subscribe CursoreLum, hasp/termo_display/state/p1b15, val; Subscribe TermostatoSet, hasp/termo_display/state/p2b1, text; Subscr..."}
14:50:33.175 RUL: Rule2 on Event#DisplayTemp do publish hasp/termo_display/command/p[1].b[20].txt %value% endon on Event#DisplayHum do publis...
14:50:33.180 RUL: Rule2 +Event#AcquaRiscSet do ot_tboiler %value% endon on Event#AcquaSanSet do ot_twater %value% endon on Event#DisplayPag d...
14:50:33.196 MQT: stat/OpenTherm1/RESULT = {"Rule2":"ON","Once":"OFF","StopOnError":"OFF","Length":785,"Free":250,"Rules":"on Event#DisplayTemp do publish hasp/termo_display/command/p[1].b[20].txt %value% endon on Event#DisplayHum do publish hasp/termo_display/command/p[1].b[21].txt %value% endon on Event#DisplayPres do publish hasp/termo_display/command/p[1].b[23].txt %value% endon on Event#PulsanteAcc do POWER1 %value% endon on Event#CursoreLum do backlog var6=100-%value%; event setvar6=1 endon  on Event#setvar6 do publish hasp/termo_display/command/dim %var6% endon on Event#TermostatoSet do mem10 %value% endon on Event#Acqua..."}
14:50:33.006 MQT: stat/OpenTherm1/RESULT = {"Rule3":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}

```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
MQT: stat/OpenTherm1/STATUS = {"Status":{"Module":0,"DeviceName":"OpenTherm1","FriendlyName":["OpenTherm1"],"Topic":"OpenTherm1","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0}}
```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
15:23:23.470 RUL: MQTT#CONNECTED performs "backlog  Subscribe DisplayTemp, hasp/termo_display/state/sensors, Temp; Subscribe DisplayHum, hasp...
15:23:23.474 SRC: Rule
15:23:23.477 CMD: Gruppo 0, Indice 1, Comando "BACKLOG", Dati "Subscribe DisplayTemp, hasp/termo_display/state/sensors, Temp; Subscribe Displ...
15:23:23.495 SRC: Backlog
15:23:23.497 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "DisplayTemp, hasp/termo_display/state/sensors, Temp"
15:23:23.501 MQT: Abbonati a hasp/termo_display/state/sensors/#
15:23:23.508 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"DISPLAYTEMP,hasp/termo_display/state/sensors/#,Temp"}
15:23:23.747 SRC: Backlog
15:23:23.750 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "DisplayHum, hasp/termo_display/state/sensors, Hum"
15:23:23.754 MQT: Abbonati a hasp/termo_display/state/sensors/#
15:23:23.767 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"DISPLAYHUM,hasp/termo_display/state/sensors/#,Hum"}
15:23:23.980 SRC: Backlog
15:23:23.983 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "DisplayPres, hasp/termo_display/state/sensors, PresSL"
15:23:23.988 MQT: Abbonati a hasp/termo_display/state/sensors/#
15:23:23.994 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"DISPLAYPRES,hasp/termo_display/state/sensors/#,PresSL"}
15:23:24.219 SRC: Backlog
15:23:24.223 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "DisplayPag, hasp/termo_display/state/page"
15:23:24.227 MQT: Abbonati a hasp/termo_display/state/page/#
15:23:24.238 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"DISPLAYPAG,hasp/termo_display/state/page/#"}
15:23:24.474 SRC: Backlog
15:23:24.478 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "DisplayDim, hasp/termo_display/state/dim"
15:23:24.482 MQT: Abbonati a hasp/termo_display/state/dim/#
15:23:24.490 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"DISPLAYDIM,hasp/termo_display/state/dim/#"}
15:23:24.724 SRC: Backlog
15:23:24.727 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "PulsanteAcc, hasp/termo_display/state/p1b14, val"
15:23:24.732 MQT: Abbonati a hasp/termo_display/state/p1b14/#
15:23:24.743 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"PULSANTEACC,hasp/termo_display/state/p1b14/#,val"}
15:23:24.974 SRC: Backlog
15:23:24.977 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "CursoreLum, hasp/termo_display/state/p1b15, val"
15:23:24.982 MQT: Abbonati a hasp/termo_display/state/p1b15/#
15:23:25.001 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"CURSORELUM,hasp/termo_display/state/p1b15/#,val"}
15:23:25.234 SRC: Backlog
15:23:25.237 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "TermostatoSet, hasp/termo_display/state/p2b1, text"
15:23:25.242 MQT: Abbonati a hasp/termo_display/state/p2b1/#
15:23:25.252 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"TERMOSTATOSET,hasp/termo_display/state/p2b1/#,text"}
15:23:25.487 SRC: Backlog
15:23:25.491 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "AcquaRiscSet, hasp/termo_display/state/p2b5, text"
15:23:25.496 MQT: Abbonati a hasp/termo_display/state/p2b5/#
15:23:25.504 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"ACQUARISCSET,hasp/termo_display/state/p2b5/#,text"}
15:23:25.743 SRC: Backlog
15:23:25.747 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "AcquaSanSet, hasp/termo_display/state/p2b3, text"
15:23:25.752 MQT: Abbonati a hasp/termo_display/state/p2b3/#
15:23:25.758 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"ACQUASANSET,hasp/termo_display/state/p2b3/#,text"}
15:23:25.993 SRC: Backlog
15:23:25.997 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "PaginaPrec, hasp/termo_display/state/p0b1, event"
15:23:25.001 MQT: Abbonati a hasp/termo_display/state/p0b1/#
15:23:26.003 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"PAGINAPREC,hasp/termo_display/state/p0b1/#,event"}
15:23:26.237 SRC: Backlog
15:23:26.240 CMD: Gruppo 0, Indice 1, Comando "SUBSCRIBE", Dati "PaginaSucc, hasp/termo_display/state/p0b3, event"
15:23:26.245 MQT: Abbonati a hasp/termo_display/state/p0b3/#
15:23:26.251 MQT: stat/OpenTherm1/RESULT = {"Subscribe":"PAGINASUCC,hasp/termo_display/state/p0b3/#,event"}
```

### TO REPRODUCE
_Steps to reproduce the behavior:_
From console:
Rule1 on mqtt#connected do backlog 
Subscribe DisplayTemp, hasp/termo_display/state/sensors, Temp;
Subscribe DisplayHum, hasp/termo_display/state/sensors, Hum; 
Subscribe DisplayPres, hasp/termo_display/state/sensors, PresSL endon

Rule2 on Event#DisplayTemp do publish hasp/termo_display/command/p[1].b[20].txt %value% endon
on Event#DisplayHum do publish hasp/termo_display/command/p[1].b[21].txt %value% endon
on Event#DisplayPres do publish hasp/termo_display/command/p[1].b[23].txt %value% endon

From MQTT publish:
hasp/termo_display/state/statusupdate {"Temp":"22.13","Hum":"30.45","Pres":"981.89","PresSL":"1027.20"}

After publishing in Tasmota console I can see:
RUL: EVENT#DISPLAYTEMP performs "publish hasp/termo_display/command/p[1].b[20].txt 22.13"
15:04:29.771 SRC: Rule
15:04:29.774 CMD: Gruppo 0, Indice 1, Comando "PUBLISH", Dati "hasp/termo_display/command/p[1].b[20].txt 22.13"
15:04:29.780 MQT: hasp/termo_display/command/p[1].b[20].txt = 22.13

When the topic is published, only one of the three events defined in rule 1 is triggered

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

I expect that in response to the subscription of the three parameters Temp, Hum, PresSL, the three events of rule2
are activated

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by lalo-uy** [comment]
I have come to this problem before.
To have 2 diferent values from a json msg you need to use 2 separated rules.

**Comment by GVollaro** [comment]
Thanks @lalo-uy for your answer but, if I understood what you are saying, I think I have already done so by defining in rule 1 an MQTT subsribe event for each of the parameters contained in the Json msg

**Comment by sfromis** [comment]
You are doing all 3 subscribes in just one rule. Try following the suggestion about separate rules, to see if you get better results then.

**Comment by GVollaro** [comment]
Even if splitting the subscriptions into different rules does not make much sense to me as if I had 4 parameters the 3 rules of tasmota would not be enough for me, I still tried to divide the subscription of the 3 parameters between rule1, rule2 and rule3 but the result was that only for the parameter defined in rule2 has the publication event been activated, for the other two nothing

**Comment by sfromis** [comment, marked as answer]
You can still have several rules in one rule set. One rule is from `on` to `endon`. Why it still does not work for you is another matter.

**Comment by GVollaro** [comment]
I raise again the problem that I have not yet been able to solve and ask me if it is not a software bug.
This is the last test I made from scratch: 
I made the latest version 9.4.0 with tasmocompiler with the following customizations:

#ifndef USE_OPENTHERM
#define USE_OPENTHERM
#endif
#define USE_EXPRESSION         
#define SUPPORT_IF_STATEMENT
#ifndef SUPPORT_MQTT_EVENT
#define SUPPORT_MQTT_EVENT
#endif

I have entered and activated the following rules:

Rule1 on mqtt#connected do backlog
Subscribe DisplayTemp, hasp/termo_display/state/sensors, Temp;
Subscribe DisplayHum, hasp/termo_display/state/sensors, Hum;
Subscribe DisplayPres, hasp/termo_display/state/sensors, PresSL endon

Rule2  on Event#DisplayTemp do publish hasp/termo_display/command/p[1].b[20].txt %value% endon
on Event#DisplayHum do publish hasp/termo_display/command/p[1].b[21].txt %value% endon
on Event#DisplayPres do publish hasp/termo_display/command/p[1].b[23].txt %value% endon

 I verified that the subscriptions had been acquired and then I published the following topic with Mqtt Expolrer:

hasp/termo_display/state/sensors = {"Temp":"23.60","Hum":"45.37","Pres":"971.64","PresSL":"1016.25"}

Result: the trigger is triggered only for the first event and the relative command is published, the events associated with the subscription of the other two parameters are not triggered.
I hope that someone can repeat the test described and help me find a solution.

**Reply by sfromis** [reply]
Possibly related, an issue has just been fixed around multiple rule triggers. Try again with the fix #11851

**Reply by GVollaro** [reply]
Thanks @sfromis,
I just recompiled the development version with the latest updates and re-run the test but the problem always appears the same way

**Reply by barbudor** [reply]
I had a very quick look at the code and my feeling is that there are multiple reasons for this case not to be supported.
For now, if you could have each sensor published on a different topic, that may work (I still have some doubts about multiple events occurring at the same time).

**Reply by GVollaro** [reply]
Thanks @barbudor,
I can try but I had to modify the source device firmware to split BME280 sensors on different topics.
From the documentation it is not clear that in the case of a multi-sensor it is possible to subscribe only one; if so I hope this limitation can be removed

**Reply by barbudor** [reply]
If the sensor device is running Tasmota, you don't need to modify the code as this can be achieved with rules.

Regarding the documentation, it only includes what users/coders have thought about. It is not that this use-case have been rejected, it's just that no-one ever thought about it and either coded or documented.
You are free to edit the documentation (pencil on top-right of the page) to insert a warning until someone can find how to handle that.

**Reply by GVollaro** [reply]
No, the sensor device is running an OpenHasp modified

**Comment by lakeweb** [comment]
I just went through this. This post was one of the hits in my searches. Just use one subscription, like:
Subscribe mqt, stat/the_lights
In your case:
Subscribe Display, hasp/termo_display/state/sensors
And, subscribe outside of the rule

Then in the rule, us the likes of:
Rule2
ON Event#mqt#power1=on DO power1 on ENDON
ON Event#mqt#power1=off DO power1 off ENDON
ON Event#mqt#power2=on DO power2 on ENDON
ON Event#mqt#power2=off DO power2 off ENDON
ON Event#mqt#power3=on DO power3 on ENDON
ON Event#mqt#power3=off DO power3 off ENDON
ON Event#mqt#power4=on DO power4 on ENDON
ON Event#mqt#power4=off DO power4 off ENDON

In your case:
Rule2
ON Event#Display#Temp DO publish hasp/termo_display/command/p[1].b[20].txt %value% ENDON
etc.

I used my test board and %value% works. That means I can turn my Rule2 into a four line rule, if I can figure out how to do a power with a logical not :)
